lib/backend/etcd/etcdv3.go

```diff
@@ -0,0 +1,452 @@
+	}
+	if config.EtcdEndpoints != "" {
+		etcdLocation = strings.Split(config.EtcdEndpoints, ",")
+	}
+
+	if len(etcdLocation) == 0 {
+		return nil, goerrors.New("no etcd authority or endpoints specified")
+	}
+
+	// Create the etcd client
+	tlsInfo := &transport.TLSInfo{
+		CAFile:   config.EtcdCACertFile,
+		CertFile: config.EtcdCertFile,
+		KeyFile:  config.EtcdKeyFile,
+	}
+	tls, _ := tlsInfo.ClientConfig()
+
+	cfg := etcdv3.Config{
+		Endpoints:   etcdLocation,
+		TLS:         tls,
+		DialTimeout: clientTimeout,
+	}
+
+	// Plumb through the username and password if both are configured.
+	if config.EtcdUsername != "" && config.EtcdPassword != "" {
+		cfg.Username = config.EtcdUsername
+		cfg.Password = config.EtcdPassword
+	}
+
+	client, err := etcdv3.New(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return &EtcdV3Client{etcdClient: client}, nil
+}
+
+// Get an entry from the datastore.  This errors if the entry does not exist.
+func (c *EtcdV3Client) Get(k model.Key) (*model.KVPair, error) {
+	key, err := model.KeyToDefaultPath(k)
+	if err != nil {
+		return nil, err
+	}
+	log.Debugf("Get Key: %s", key)
+	resp, err := c.etcdClient.Get(context.Background(), key)
+	if err != nil {
+		return nil, errors.ErrorDatastoreError{Err: err}
+	}
+
+	// Older deployments with etcd may not have the Host metadata, so in the
+	// event that the key does not exist, just do a get on the directory to
+	// check it exists, and if so return an empty Metadata.
+	if resp.Count == 0 {
+		if _, ok := k.(model.HostMetadataKey); ok {
+			return c.getHostMetadataFromDirectory(k)
+		}
+
+		return nil, errors.ErrorResourceDoesNotExist{Identifier: k}
+	}
+
+	kv := resp.Kvs[0]
+
+	v, err := model.ParseValue(k, []byte(kv.Value))
+	if err != nil {
+		return nil, err
+	}
+
+	return &model.KVPair{Key: k, Value: v, Revision: kv.ModRevision}, nil
+}
+
+// Create an entry in the datastore.  This errors if the entry already exists.
+func (c *EtcdV3Client) Create(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	return c.create(d, logCxt)
+}
+
+// Update an entry in the datastore.  This errors if the entry already exists.
+func (c *EtcdV3Client) Update(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	return c.update(d, logCxt)
+}
+
+// Apply an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *EtcdV3Client) Apply(d *model.KVPair) (*model.KVPair, error) {
+	logCxt := log.WithFields(log.Fields{"key": d.Key, "value": d.Value, "ttl": d.TTL, "rev": d.Revision})
+	return c.apply(d, logCxt)
+}
+
+// Delete an entry in the datastore.  This errors if the entry does not exists.
+func (c *EtcdV3Client) Delete(d *model.KVPair) error {
+	key, err := model.KeyToDefaultDeletePath(d.Key)
```

Comment:
Unless it's handled somewhere else that I didn't spot, this is bugged.  Since etcdv3 doesn't have a concept of a directory, deleting `foo/bar` may delete `foo/barrel` too.